eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala

```diff
@@ -274,28 +288,45 @@ object Graph {
-    val cltvFactor = normalize(edge.update.cltvExpiryDelta.toInt, CLTV_LOW, CLTV_HIGH)
-
-    // NB we're guaranteed to have weightRatios and factors > 0
-    val factor = weightRatios.baseFactor + (cltvFactor * weightRatios.cltvDeltaFactor) + (ageFactor * weightRatios.ageFactor) + (capFactor * weightRatios.capacityFactor)
-    val totalWeight = prev.weight + (fee + hopCost).toLong * factor
-    RichWeight(totalCost, prev.length + 1, totalCltv, totalWeight)
+  private def addEdgeWeight(sender: PublicKey, edge: GraphEdge, prev: RichWeight, currentBlockHeight: Long, weightRatios: Either[WeightRatios, HeuristicsConstants], includeLocalChannelCost: Boolean): RichWeight = {
+    val totalAmount = if (edge.desc.a == sender && !includeLocalChannelCost) prev.amount else addEdgeFees(edge, prev.amount)
+    val fee = totalAmount - prev.amount
+    val totalFees = prev.fees + fee
+    val cltv = if (edge.desc.a == sender && !includeLocalChannelCost) CltvExpiryDelta(0) else edge.update.cltvExpiryDelta
+    val totalCltv = prev.cltv + cltv
+    weightRatios match {
+      case Left(weightRatios) =>
+        val hopCost = nodeFee(weightRatios.hopCost.feeBase, weightRatios.hopCost.feeProportionalMillionths, prev.amount)
+        import RoutingHeuristics._
+
+        // Every edge is weighted by funding block height where older blocks add less weight. The window considered is 1 year.
+        val channelBlockHeight = ShortChannelId.coordinates(edge.desc.shortChannelId).blockHeight
+        val ageFactor = normalize(channelBlockHeight, min = (currentBlockHeight - BLOCK_TIME_ONE_YEAR).toDouble, max = currentBlockHeight.toDouble)
+
+        // Every edge is weighted by channel capacity, larger channels add less weight
+        val edgeMaxCapacity = edge.capacity.toMilliSatoshi
+        val capFactor = 1 - normalize(edgeMaxCapacity.toLong.toDouble, CAPACITY_CHANNEL_LOW.toLong.toDouble, CAPACITY_CHANNEL_HIGH.toLong.toDouble)
+
+        // Every edge is weighted by its cltv-delta value, normalized
+        val cltvFactor = normalize(edge.update.cltvExpiryDelta.toInt, CLTV_LOW, CLTV_HIGH)
+
+        // NB we're guaranteed to have weightRatios and factors > 0
+        val factor = weightRatios.baseFactor + (cltvFactor * weightRatios.cltvDeltaFactor) + (ageFactor * weightRatios.ageFactor) + (capFactor * weightRatios.capacityFactor)
+        val totalWeight = prev.weight + (fee + hopCost).toLong * factor
+        RichWeight(totalAmount, prev.length + 1, totalCltv, 1.0, totalFees, totalWeight)
+      case Right(heuristicsConstants) =>
+        val hopCost = nodeFee(heuristicsConstants.hopCost.feeBase, heuristicsConstants.hopCost.feeProportionalMillionths, prev.amount)
+        val cltv = if (edge.desc.a == sender && !includeLocalChannelCost) CltvExpiryDelta(0) else edge.update.cltvExpiryDelta
+        val totalCltv = prev.cltv + cltv
+        val riskCost = totalAmount.toLong * cltv.toInt * heuristicsConstants.lockedFundsRisk
```

Comment:
That will very strongly bias towards smaller routes, because for each hop you're re-applying the risk to the total amount.
This makes sense to me (I do think this is how it should be computed) but then the additional `hopCost` mechanism is redundant, isn't it?

We could simplify this heuristic by only applying this funds-locked-risk and nothing else, wouldn't that be better?

Reply:
> Actually I don't think it will. It would bias toward shorter routes if all channels had a similar cltv delta but that's not the case, both 40 and 144 are common values. The riskCost will be lower for a route of length 3 with cltv deltas of 40 than a route of size 1 with a cltv delta of 144.

Good point, it's true that we're currently seeing widely different values of cltv-expiry-delta, so this isn't a proxy heuristic for route length. I'm convinced, we should keep `hopCost` here as well!